masu/database/provider_status_accessor.py

```diff
@@ -0,0 +1,212 @@
+
+        Args:
+            None
+        Returns:
+            (datetime): Timestamp for the status
+        """
+        return self._obj.timestamp
+
+    def get_retries(self):
+        """
+        Return the provider status retry count.
+
+        Args:
+            None
+        Returns:
+            (int): Status retry count
+        """
+        return self._obj.retries
+
+    def is_backing_off(self):
+        """Determine if the provider is waiting to retry."""
+        def backoff(interval, maximum=64):
+            """Exponential back-off."""
+            # (12 + interval) is used to select desirable values from the
+            # exponential progression.
+            return min(maximum, (2 ** (12 + interval))) + (random.randint(0, 1000) / 1000.0)
+
+        retries = self.get_retries()
+        timestamp = self.get_timestamp()
+        backoff_threshold = timestamp + timedelta(seconds=backoff(retries,
+                                                                  maximum=86400))
+        if self.get_status() == ProviderStatusCode.WARNING and \
+                retries <= ProviderStatusAccessor.MAX_RETRIES and \
+                datetime.now(timezone.utc) <= backoff_threshold:
+            return True
+        return False
+
+    def is_valid(self):
+        """Determine if the provider is valid based on its previous status."""
+        if self.get_status() == ProviderStatusCode.READY:
+            return True
+
+        if self.get_status() == ProviderStatusCode.WARNING and \
+                self.get_retries() <= ProviderStatusAccessor.MAX_RETRIES:
+            return True
+
+        return False
+
+    def update_status(self, status, error=None):
+        """Update the provider status.
+
+        Args:
+            status (int) Provider status code
+            error (Exception) Exception raised during processing attempts
+
+        Returns:
+            None
+
+        """
+        self._obj.status = status
+        self._obj.timestamp = datetime.now(timezone.utc)
+
+        if status == ProviderStatusCode.READY:
+            self._obj.retries = 0    # reset counter on success
+        elif status == ProviderStatusCode.WARNING:
+            self._obj.retries += 1
+
+        if error is not None:
+            self._obj.last_message = str(error)
+        else:
+            self._obj.last_message = 'none'
+
+        self.commit()
+
+    def update_status_error(self, error=None):
+        """Update status to an error state.
+
+        Args:
+            error (Exception) Exception raised during processing attempts
+
+        """
+        # stay in warning state until max retries is exceeded
+        if self.get_retries() >= ProviderStatusAccessor.MAX_RETRIES:
+            self.update_status(ProviderStatusCode.DISABLED_ERROR, error)
+
+        # stay in error state until Provider is fixed.
+        # state will be reset by provider updates made via Koku API
+        if self.get_status() not in [ProviderStatusCode.DISABLED_ERROR,
```

Comment:
As mentioned earlier, I think we need to make this type of change in a separate PR once we get commitment on API/UI changes.  We don't want to ever lock out an account with no way of unlocking.